homeassistant/components/zwave_mqtt/entity.py

```diff
@@ -0,0 +1,272 @@
+
+from . import const
+from .const import DOMAIN, PLATFORMS
+from .discovery import check_node_schema, check_value_schema
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ZWaveDeviceEntityValues:
+    """Manages entity access to the underlying Z-Wave value objects."""
+
+    def __init__(self, hass, options, schema, primary_value):
+        """Initialize the values object with the passed entity schema."""
+        self._hass = hass
+        self._entity_created = False
+        self._schema = copy.deepcopy(schema)
+        self._values = {}
+        self.options = options
+
+        # Go through values listed in the discovery schema, initialize them,
+        # and add a check to the schema to make sure the Instance matches.
+        for name, disc_settings in self._schema[const.DISC_VALUES].items():
+            self._values[name] = None
+            disc_settings[const.DISC_INSTANCE] = [primary_value.instance]
+
+        self._values[const.DISC_PRIMARY] = primary_value
+        self._node = primary_value.node
+        self._schema[const.DISC_NODE_ID] = [self._node.node_id]
+
+    def setup(self):
+        """Set up values instance."""
+        # Check values that have already been discovered for node
+        # and see if they match the schema and need added to the entity.
+        for value in self._node.values():
+            self.check_value(value)
+
+        # Check if all the _required_ values in the schema are present and
+        # create the entity.
+        self._check_entity_ready()
+
+    def __getattr__(self, name):
+        """Get the specified value for this entity."""
+        return self._values.get(name, None)
+
+    def __iter__(self):
+        """Allow iteration over all values."""
+        return iter(self._values.values())
+
+    def __contains__(self, name):
+        """Check if the specified name/key exists in the values."""
+        return name in self._values
+
+    @callback
+    def check_value(self, value):
+        """Check if the new value matches a missing value for this entity.
+
+        If a match is found, it is added to the values mapping.
+        """
+        # Make sure the node matches the schema for this entity.
+        if not check_node_schema(value.node, self._schema):
+            return
+
+        # Go through the possible values for this entity defined by the schema.
+        for name in self._values:
+            # Skip if it's already been added.
+            if self._values[name] is not None:
+                continue
+            # Skip if the value doesn't match the schema.
+            if not check_value_schema(value, self._schema[const.DISC_VALUES][name]):
+                continue
+
+            # Add value to mapping.
+            self._values[name] = value
+
+            # If the entity has already been created, notify it of the new value.
+            if self._entity_created:
+                async_dispatcher_send(self._hass, f"{self.values_id}_value_added")
```

Comment:
All dispatcher signals need to be prefixed with `zwave_`

Reply:
Commit done.